### PR TITLE
Cleanup error handling in commissioner

### PR DESCRIPTION
**Pull request by cecille**
#### Problem
- Error handling was inconsistent in commissioning due to PRs merging at different times
- failures during op cert provisioning and CASE session establishment didn't properly clean up
- PASE session is prematurely removed before CASE is established
- CASE manager wasn't notified on CASE failure

#### Change overview
- General cleanup of error handling function calls in commissioning
- Changed op cert callbacks to notify the auto commissioner rather than calling OnSessionEstablishmentError directly - the session establishment error function is for PASE
- AutoCommissioner now calls kCleanup on error
- PASE cleanup moved to after CASE is established
- CASE manager session is cleaned up on CASE failure

#### Testing
Forced a single commissioning failure on lighting app to simulate a transient failure - tested with failure when adding NOC and a failure when establishing CASE. Tested with chip-device-ctrl by first establishing a pase session, then running commission. The first commission call fails (as expected), second commissioning call works for both the AddNOC failure and the CASE failure.

### CODE REVIEW

**Review by woody-apple — APPROVED** [review]

**Review by bzbarsky-apple — APPROVED** [review]

File: src/controller/AutoCommissioner.cpp [hunk 1]
```diff
@@ -249,56 +249,62 @@ CHIP_ERROR AutoCommissioner::CommissioningStepFinished(CHIP_ERROR err, Commissio
...
+            SetPAI(report.Get<RequestedCertificate>().certificate);
+            break;
+        case CommissioningStage::kSendDACCertificateRequest:
+            SetDAC(report.Get<RequestedCertificate>().certificate);
+            break;
+        case CommissioningStage::kSendAttestationRequest:
+            // These don't need to be deep copied to local memory because they are used in this one step then never again.
+            mParams.SetAttestationElements(report.Get<AttestationResponse>().attestationElements)
+                .SetAttestationSignature(report.Get<AttestationResponse>().signature);
+            // TODO: Does this need to be done at runtime? Seems like this could be done earlier and we woouldn't need to hold a
```
Comment (andy31415) [inline src/controller/AutoCommissioner.cpp:267]: nit: extra "o" in wooldn't. Not new code though

**Review by andy31415 — APPROVED** [review]

File: src/controller/AutoCommissioner.cpp [hunk 2]
```diff
@@ -249,56 +249,62 @@ CHIP_ERROR AutoCommissioner::CommissioningStepFinished(CHIP_ERROR err, Commissio
...
+        case CommissioningStage::kSendDACCertificateRequest:
+            SetDAC(report.Get<RequestedCertificate>().certificate);
+            break;
+        case CommissioningStage::kSendAttestationRequest:
+            // These don't need to be deep copied to local memory because they are used in this one step then never again.
+            mParams.SetAttestationElements(report.Get<AttestationResponse>().attestationElements)
+                .SetAttestationSignature(report.Get<AttestationResponse>().signature);
+            // TODO: Does this need to be done at runtime? Seems like this could be done earlier and we wouldn't need to hold a
+            // reference to the operational credential delegate here
+            if (mOperationalCredentialsDelegate != nullptr)
```
Comment (tcarmelveilleux) [inline src/controller/AutoCommissioner.cpp:269]: Yes, it needs to be done at runtime

File: src/controller/CHIPDeviceController.cpp [hunk 3]
```diff
@@ -1475,20 +1474,25 @@ void BasicFailure(void * context, uint8_t status)
...
+    report.stageCompleted = mCommissioningStage;
+    CHIP_ERROR status     = mCommissioningDelegate->CommissioningStepFinished(err, report);
     if (status != CHIP_NO_ERROR)
     {
-        OnSessionEstablishmentError(status);
+        // Commissioning delegate will only return error if it failed to perform the appropriate commissioning step.
+        // In this case, we should call back the commissioning complete and call session error
+        if (mPairingDelegate != nullptr)
+        {
+            mPairingDelegate->OnCommissioningComplete(mDeviceBeingCommissioned->GetDeviceId(), status);
```
Comment (tcarmelveilleux) [inline src/controller/CHIPDeviceController.cpp:1494]: OnCommissioningComplete drops the NodeID to the floor. It seems that we should provide a full operational ID on OnCommissioningComplete, not just NodeID, if we pass anything. Right now, all in-SDK users disregard the NodeID, adn a NodeID without Root public key and fabric ID is ambiguous, so we should probably avoid passing it.

**Review by tcarmelveilleux — APPROVED** [review]

File: src/controller/CHIPDeviceController.cpp (hunk 3 above)
Comment (cecille) [reply src/controller/CHIPDeviceController.cpp:1494]: Fair enough - can address in a follow up. When this was written, node id was king. https://github.com/project-chip/connectedhomeip/issues/14357

File: src/controller/AutoCommissioner.cpp (hunk 2 above)
Comment (cecille) [reply src/controller/AutoCommissioner.cpp:269]: okie dokie. Will remove the todo in a follow up (don't want to wait for the CI again)

**Review by woody-apple — APPROVED** [review]

**Review by msandstedt — APPROVED** [review]

**Review by woody-apple — APPROVED** [review]

**Pull request merged**